File: 572/subtree-of-another-tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} s
 * @param {TreeNode} t
 * @return {boolean}
 */
var isSubtree = function (s, t) {
    const isSameTree = function (p, q) {
        if (p != null && q != null) {
            if (p.val == q.val)
                return isSameTree(p.left, q.left) && isSameTree(p.right, q.right);
            else return false;
        } else if (p == null && q == null) {
            return true;
        } else return false;
    };

    if (s == null && t == null)
        return true;
    else {
        if (isSameTree(s, t))
            return true;
        else if (s.left != null && s.right != null)
            return isSubtree(s.left, t) || isSubtree(s.right, t);
        else if (s.left != null)
            return isSubtree(s.left, t);
        else if (s.right != null)
            return isSubtree(s.right, t);
        else
            return false;
    }
};

File: 572/subtree-of-another-tree.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func isSubtree(s *TreeNode, t *TreeNode) bool {
	if s == nil && t == nil {
		return true
	} else {
		if isSameTree(s, t) {
			return true
		} else if s.Left != nil && s.Right != nil {
			return isSubtree(s.Left, t) || isSubtree(s.Right, t)
		} else if s.Left != nil {
			return isSubtree(s.Left, t)
		} else if s.Right != nil {
			return isSubtree(s.Right, t)
		} else {
			return false
		}
	}
}

func isSameTree(p *TreeNode, q *TreeNode) bool {
	if p != nil && q != nil {
		if p.Val == q.Val {
			return isSameTree(p.Left, q.Left) && isSameTree(p.Right, q.Right)
		} else {
			return false
		}
	} else if p == nil && q == nil {
		return true
	} else {
		return false
	}
}